src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs

```diff
@@ -235,40 +233,231 @@ internal void WriteTopLevelProperty(ODataProperty property)
+                    ODataPayloadKind kind = GetPayloadKind();
+
+                    if (!(this.JsonLightOutputContext.MetadataLevel is JsonNoMetadataLevel))
+                    {
+                        ODataContextUrlInfo contextInfo = GetContextUrlInfo(property);
+                        await this.WriteContextUriPropertyAsync(kind, () => contextInfo)
+                            .ConfigureAwait(false);
+                    }
+
+                    // Note we do not allow named stream properties to be written as top level property.
+                    this.JsonLightValueSerializer.AssertRecursionDepthIsZero();
+                    await this.WritePropertyAsync(
+                        property,
+                        null /*owningType*/,
+                        true /* isTopLevel */,
+                        this.CreateDuplicatePropertyNameChecker(),
+                        null /* metadataBuilder */).ConfigureAwait(false);
+                    this.JsonLightValueSerializer.AssertRecursionDepthIsZero();
+
+                    await this.AsynchronousJsonWriter.EndObjectScopeAsync().ConfigureAwait(false);
+                });
+        }
+
+        /// <summary>
+        /// Asynchronously writes property names and value pairs.
+        /// </summary>
+        /// <param name="owningType">The <see cref="IEdmStructuredType"/> of the resource (or null if not metadata is available).</param>
+        /// <param name="properties">The enumeration of properties to write out.</param>
+        /// <param name="isComplexValue">
+        /// Whether the properties are being written for complex value. Also used for detecting whether stream properties
+        /// are allowed as named stream properties should only be defined on ODataResource instances
+        /// </param>
+        /// <param name="duplicatePropertyNameChecker">The DuplicatePropertyNameChecker to use.</param>
+        /// <param name="metadataBuilder">The metadatabuilder for writing the property.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal async Task WritePropertiesAsync(
+            IEdmStructuredType owningType,
+            IEnumerable<ODataProperty> properties,
+            bool isComplexValue,
+            IDuplicatePropertyNameChecker duplicatePropertyNameChecker,
+            ODataResourceMetadataBuilder metadataBuilder)
+        {
+            if (properties == null)
             {
-                this.currentPropertyInfo = new PropertySerializationInfo(this.JsonLightOutputContext.Model, propertyName, owningType) { IsTopLevel = isTopLevel };
+                return;
             }
-            else
+
+            foreach (ODataProperty property in properties)
             {
-                this.currentPropertyInfo = this.JsonLightOutputContext.PropertyCacheHandler.GetProperty(this.JsonLightOutputContext.Model, propertyName, owningType);
+                await this.WritePropertyAsync(

```

Comment:
@marabooy It does change the semantics. Since we're writing to a single stream, we need the first write operation to complete before the second commence